internal/controller/state/graph/route_common.go

```diff
@@ -864,38 +812,31 @@ func tryToAttachL7RouteToListeners(
 		return conditions.NewRouteNoMatchingListenerHostname(), false
 	}
 
-	if !attachedToAtLeastOneValidListener {
+	if !attachableListener.Valid {
 		return conditions.NewRouteInvalidListener(), true
 	}
 
 	return conditions.Condition{}, true
 }
 
-// findAttachableListeners returns a list of attachable listeners and whether the listener exists for a non-empty
-// sectionName.
-func findAttachableListeners(sectionName string, listeners []*Listener) ([]*Listener, bool) {
-	if sectionName != "" {
-		for _, l := range listeners {
-			if l.Name == sectionName {
-				if l.Attachable {
-					return []*Listener{l}, true
-				}
-				return nil, true
-			}
-		}
-		return nil, false
-	}
+// findAttachableListener returns the attachable listener that matches the sectionName and port.
+// Since buildSectionNameRefs always expands to concrete sectionName+port combinations,
+// we only need to handle the sectionName case.
+func findAttachableListener(ref *ParentRef, listeners []*Listener) (*Listener, bool) {
+	sectionName := string(*ref.SectionName)
 
-	attachableListeners := make([]*Listener, 0, len(listeners))
 	for _, l := range listeners {
-		if !l.Attachable {
-			continue
+		if l.Name == sectionName {
+			if l.Attachable && (ref.Port == nil || l.Source.Port == *ref.Port) {
+				return l, true
+			}
+			if ref.Port != nil && l.Source.Port != *ref.Port {
+				return nil, false
+			}
```

Comment:
Would this need to be, since the listener does match a section name? Also can we add back a description of what the boolean does to the function comment?

From the check above:

```
// Case 1: Attachment is not possible because the specified SectionName does not match any Listeners in the
// Gateway.
```

```suggestion
			if ref.Port != nil && l.Source.Port != *ref.Port {
				return nil, true
			}
```

Reply:
From what I understand the boolean it seems sets a condition farther downstream in the code. If I have it set to true, then the condition will be like invalidListener because the listener was found and wasn't attachable, but the port filter works differently as the listener was not ever found in the first place I guess when the ports don't match so that condition has to be NoMatchingParent which means the boolean must be true. 

sorry for the long answer hopefully its not hard to follow this was actually the if clause I need that was failing the conformance tests originally